smtp.c

```diff
@@ -14,8 +15,13 @@ static void get_random_bytes(void *buf, size_t n)
 {
 #if defined HAVE_ARC4RANDOM  /* BSD */
 	arc4random_buf(buf, n);
-#elif defined HAVE_GETRANDOM /* Linux */
+#elif defined HAVE_GETRANDOM /* Linux, GLIBC >= 2.25 */
 	getrandom(buf, n, 0);
+#elif defined __linux__ && defined __GLIBC__ && \
+    __GLIBC__ <= 2 &&  __GLIBC_MINOR__ < 25 /* Linux, GLIBC < 2.25 */
```

Comment:
Rather than probing for certain library versions, what about testing directly whether `syscall(SYS_getrandom ...)` compiles? The configure script has examples -- you create a sample block of code and pass it to the `compiles` function. You could have the configure script set HAVE_GETRANDOM_SYSCALL in this case, then check for that feature macro here.

Reply:
That's a great idea indeed, that would be much cleaner.

I'm not too sure how to check for a syscall existence at compilation time, though, without executing the resulting code: the examples in the `configure` script are just testing the return code of the compiler, so to keep the same spirit, the compiler would have to throw an error if the `SYS_getrandom` syscall doesn't exist.

Do you have any suggestion?

Reply:
I'm thinking we test whether syscall and SYS_getrandom are defined and have the expected type. If we can assign the function to a function pointer then it must exist. Similarly we attempt to assign the particular sys call code to a variable of the expected type. We use those values in the return code so that the compiler doesn't warn that they're unused.

```sh
if compiles "-D_GNU_SOURCE" "
	#include <unistd.h>
	#include <sys/syscall.h>
	int main(void)
	{
		long n = SYS_getrandom;
		long (*p)(long, ...) = syscall;
		return (intptr_t)p & n;
	}"
then
	echo "CFLAGS += -DHAVE_GETRANDOM_SYSCALL" >> config.mk
fi
```

You may need to make modifications, I haven't tested this.

Reply:
Oh yeah, that's much better, thank you!